### PR TITLE
Sort properties and defaults when encoding PDLs.

### PR DESCRIPTION
**Pull request by tjni**
The PDL encoder already sorts the top-level properties on a field by
their names. This extends this sorting to nested properties.

An example of a nested property that is sorted is:

    @nested = {
      "a" : "a",
      "b" : "b"
    }

When default values are records, their fields are sorted in the order
that the fields are defined in the schema. For example:

    date: record Date {
      month: int
      day: int
      year: int
    } = {
      "month": 1,
      "day": 1,
      "year": 2021
    }

This enables the PDL encoder to be used for autoformatting PDLs, since
the order of map entries will be deterministic.

### CODE REVIEW

**Comment by junchuanwang** [comment]
Can you also test if this makes other things sorted when encoding pdl, such as field default value list(or map)?

**Comment by tjni** [comment]
It does appear to also sort the default value when it's a map. Maybe this shouldn't happen, since the fields of the record corresponding to the default value might not be sorted?

**Comment by junchuanwang** [comment]
Sorted order in default value actually seems OK to me but I will confirm with the team

**Comment by evanw555** [comment]
Synced up with @tjni offline. The requirement is just for the JSON objects in the PDL properties to be encoded deterministically. We've determined that sorting them alphabetically is fine, as users probably won't care enough to preserve the original ordering of a JSON object...

As for the default value JSON, we determined that the ideal behavior here is to sort the JSON objects in the record's original field ordering. This would require the JSON encoder to be schema-aware. @tjni is currently looking into this.

**Comment by aman1309** [comment]
> Synced up with @tjni offline. The requirement is just for the JSON objects in the PDL properties to be encoded deterministically. We've determined that sorting them alphabetically is fine, as users probably won't care enough to preserve the original ordering of a JSON object...
> 
> As for the default value JSON, we determined that the ideal behavior here is to sort the JSON objects in the record's original field ordering. This would require the JSON encoder to be schema-aware. @tjni is currently looking into this.

we could have two Json encoder one for properties and another for others. The Codec can not be schema aware.

**Comment by tjni** [comment]
I've updated the code to use a different codec for schema-aware sorting of JSON maps. I found `JacksonDataTemplateCodec`, which I am reusing for this.

File: data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java
Comment (junchuanwang) [inline data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java:137]: can you format this as well, thanks!

File: data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java
Comment (junchuanwang) [inline data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java:55]: Can you do one more test on nested properties larger than 1 nested layer?

**Review by junchuanwang — COMMENTED** [review]
No other concerns if the comments of the tests can be fixed.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PdlBuilder.java:264]: I recommend not introducing a method unique to default values, but rather just extend the `writeJson` method to accept a second parameter (of type `DataSchema`). For this parameter, a non-null value indicates a schema-based ordering to use, and `null` indicates to just use alphabetical ordering. This would be cleaner IMO.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PdlBuilder.java:262]: If you agree with my suggestion below, then this method should no longer be abstract and it instead should just delegate to `writeJson(value, null)`

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PdlBuilder.java:56]: This might cause confusion, since the compact PDL builder has its own concept of a "buffer". I suggest renaming to `DEFAULT_JSON_CODEC_BUFFER` or something.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PdlBuilder.java:287]: Similar to my other suggestion, I suggest collapsing `toOrderedBySchemaJson` and `toJson` into one unified interface which accepts a `DataSchema` that may be null or non-null (e.g. `toJson(value, jsonCodec, schema)`. For all these methods, the javadoc should clearly state how the ordering is affected by null vs non-null arguments.

**Review by evanw555 — COMMENTED** [review]
Thanks for the quick changes. Lucky that we already have a schema-aware JSON codec ;)

My comments are generally about keeping the interfaces in these classes clean by avoiding the creation of too many specific methods.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (evanw555) [reply data/src/main/java/com/linkedin/data/schema/PdlBuilder.java:264]: If you make this change, document this clearly in the javadoc

File: data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java
Comment (tjni) [reply data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java]: I added indentation on each line and trim it prior to joining each string. I hope it looks better now.

File: data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java
Comment (tjni) [reply data/src/test/java/com/linkedin/data/schema/TestSchemaToPdlEncoder.java:57]: Good idea! Added testEncodeSortsMultiLevelNestedPropertyMap.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (tjni) [reply data/src/main/java/com/linkedin/data/schema/PdlBuilder.java]: That makes sense to me.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (tjni) [reply data/src/main/java/com/linkedin/data/schema/PdlBuilder.java]: Done!

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (tjni) [reply data/src/main/java/com/linkedin/data/schema/PdlBuilder.java]: This isn't possible at the moment because the codec that both each code path accepts is different, and there does not appear to be a common method I can call. I've renamed `toOrderedBySchemaJson` to `toJson` though and added documentation for the behavior.

It might be possible to refactor the underlying codecs so that the PdlBuilder can call the same method on both, which would make this possible. Let me know if you think it's worth pursuing.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java:70]: Should've commenting on this earlier, but this should be renamed to `_sortKeys`, since "properties" aren't a concept that's inherently known by the JSON codec.

Also, just making sure... this doesn't sort array values as well, right?

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PdlBuilder.java:328]: This naming isn't quite accurate. It's technically not transforming the data "to a primitive", it's transforming "from primitive data to primitive JSON data".

However, I think dividing the logic like this isn't totally necessary. I suggest packing this logic back into `toJson(value, codec)` and having `toJson(value, codec, schema)` just delegate back to that one for the primitive case.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java
Comment (tjni) [reply data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java:70]: Good catch. It should be `_sortKeys` and it doesn't sort arrays. Let me make sure I check that in the unit test.

File: data/src/main/java/com/linkedin/data/schema/PdlBuilder.java
Comment (tjni) [reply data/src/main/java/com/linkedin/data/schema/PdlBuilder.java]: I missed the fact that `JacksonDataTemplateCodec` extends `JacksonDataCodec`, which makes this possible. Doing this now.

**Review by evanw555 — APPROVED** [review]
LGTM! Thanks for the changes. Let's wait until tomorrow to merge this, just to let others who don't work at night leave comments (and to potentially bundle other changes in with this minor version).

**Review by junchuanwang — APPROVED** [review]

**Pull request merged**
